sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala

```diff
@@ -441,25 +428,76 @@ trait ProverInterpreter extends Interpreter with AttributionCore {
       t.withPolynomial(q).copy(children = newChildren)
 
     // If the node is a leaf marked "real", compute its response according to the second prover step
-    // of the Sigma-protocol given the commitment, challenge, and witness
+    // of the Sigma-protocol given the commitment, challenge, and witness, or pull response from the hints bag
     case su: UnprovenSchnorr if su.real =>
-      assert(su.challengeOpt.isDefined, s"Real UnprovenTree $su should have challenge defined")
-      val privKey = secrets
+      assert(su.challengeOpt.isDefined, s"Real UnprovenSchnorr $su should have challenge defined")
+      val privKeyOpt = secrets
         .filter(_.isInstanceOf[DLogProverInput])
         .find(_.asInstanceOf[DLogProverInput].publicImage == su.proposition)
-        .get.asInstanceOf[DLogProverInput]
-      val z = DLogInteractiveProver.secondMessage(privKey, su.randomnessOpt.get, su.challengeOpt.get)
+
+      val z = privKeyOpt match {
+        case Some(privKey) =>
```

Comment:
Wouldn't `case Some(privKey:DLogProverInput)` be better to avoid typecasts?

Reply:
Yes! done